### PR TITLE
Add lint `[bool_to_int_with_if]`

### LINKED ISSUE #8131: Suggest casting bool to int instead of using if statement.
### What it does

Instead of using an if statement to convert a bool to an int, this lint suggests using a coercion or a `from()` function.

### Lint Name

bool_to_int_with_if

### Category

style

### Advantage

- Makes code more concise
- Makes this conversion more discoverable

### Drawbacks

- None

### Example

```rust
if condition {
    1_i64
} else {
    0
}
```

Could be written as:

```rust
i64::from(condition)
```

or

```rust
condition as i64
```

### PR DESCRIPTION
**Pull request by jst-r**
If you added a new lint, here's a checklist for things that will be
checked during review or continuous integration.

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

---

fixes #8131

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: added lint ``[`bool_to_int_with_if`]``

### CODE REVIEW

File: tests/ui/author/struct.rs
Comment (xFrednet) [inline tests/ui/author/struct.rs:10]: Haven't looked at the PR in detail yet. Could you revert this and allow the new lint in this file? (Or alternatively just use different number literals?)

**Comment by xFrednet** [comment]
This looks like a fun one, and I've already had a discussion in the issue. I'm claiming the review, I hope that's okay.

r? @xFrednet

I'll do a full review over the weekend.

@jst-r The commits use a different username and email than your account. This is totally fine for us, but I wanted to point that out, in case that was a mistake :upside_down_face:

**Comment by jst-r** [comment]
@xFrednet I'm aware of different account. It's my gitlab account, I haven't changed credentials in my terminal. Glad this isn't an issue.

File: tests/ui/author/struct.rs
Comment (jst-r) [reply tests/ui/author/struct.rs]: Fixed in [26468a8](https://github.com/rust-lang/rust-clippy/pull/9086/commits/26468a814936662a14ea1500dea4ba970def0646)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:51]: NIT: If chains are usually used with curly brackets. Alternatively, you can also use rust's if let chains. They cover the same functionality and allow automatic formatting.

See: https://rust-lang.github.io/rfcs/2497-if-let-chains.html

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:65]: Here, I would either store `lbrace` and `rbrace` in a value, or change the formatting as the line is quite long. rustfmt sadly doesn't work inside if_chains :)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:64]: There are cases, where these snippet functions can't get the correct snippet, in those cases the placeholder is returned. Can you use [`snippet_block_with_applicability`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/source/fn.snippet_block_with_applicability.html) to ensure that the applicability is also adjusted correctly? The code would looks something like this:

```rs
			let mut app = Applicability::MachineApplicable;
            let snippet = snippet_block_with_applicability(ctx, check.span, "..", None, &mut app);
```

Then you can later use the applicability in `diag.span_suggestion(...)`

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:93]: You can also leave these out completely and replace them with `..`. That's usually better, if you only need a few values.

```suggestion
            ..
```

**Review by xFrednet — COMMENTED** [review]
Overall looks good to me, the direction is already correct, and it even found some improvements in Clippy which is always nice! Most of my comments are small and should be simple to fix. :)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [reply clippy_lints/src/bool_to_int_with_if.rs:51]: Same goes for other `if_chains!` macros,  I just didn't add any comments for them :upside_down_face:

File: tests/ui/bool_to_int_with_if.rs
Comment (jst-r) [reply tests/ui/bool_to_int_with_if.rs:52]: Thanks for suggesting test cases.
Turns out second case produces invalid code. I'll fix it and think more carefully about checking parent node.

Feels like suggesting wrapping it in else is better. And since I'll have to work on that case anyway, it shouldn't take to much extra time to implement.

Also I think suggesting inversion in cases like this would help with consistency.

```rust
if a {
    0
} else {
    1
}

// suggest
usize::from(!a)
```

I can feel myself getting carried away a little bit. But it's a lot fun, and imo it's better to release a good lint later than a half-baked sooner.

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (jst-r) [reply clippy_lints/src/bool_to_int_with_if.rs]: Didn't know about that. May be that could be a lint? Seems like it wouldn't be too hard to implement.

File: tests/ui/bool_to_int_with_if.rs
Comment (xFrednet) [reply tests/ui/bool_to_int_with_if.rs:52]: > Turns out second case produces invalid code. I'll fix it and think more carefully about checking parent node.

You might want to take a look at [`is_else_clause`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.is_else_clause.html). That should do the check for you.

> I can feel myself getting carried away a little bit. But it's a lot fun, and imo it's better to release a good lint later than a half-baked sooner.

Getting carried away is a problem, I also have. The good thing is, that that is totally fine here. That's also a thing I enjoy about this and other open source projects. The focus is on the community and not on any deadline. So feel free to take your time. :upside_down_face: Alternatively, we can also get this merge ready, and you can improve the lint later :)

**Comment by xFrednet** [comment]
Hey @jst-r, this is a ping from triage. Are you stuck somewhere or need more help with the added comments?

@rustbot author

**Comment by jst-r** [comment]
#

@xFrednet
Everything is pretty clear. Sorry for going silent. I'll finish working on your suggestions this week. If I get stuck on anything I'll let you know.

> Hey @jst-r, this is a ping from triage. Are you stuck somewhere or need more help with the added comments?

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (jst-r) [reply clippy_lints/src/bool_to_int_with_if.rs]: fixed in [d857ce3](https://github.com/rust-lang/rust-clippy/pull/9086/commits/d857ce3109209ec6c16a9515821675a7e1cd211e)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (jst-r) [reply clippy_lints/src/bool_to_int_with_if.rs]: fixed in [d857ce3](https://github.com/rust-lang/rust-clippy/pull/9086/commits/d857ce3109209ec6c16a9515821675a7e1cd211e).

I changed all `if_chain!` to rust's if chains.

Though formatting still doesn't seem to work. It may be an issue with rustfmt. After looking through open issues in its repo I didn't find any relating to `if let` chains specifically.

And I feel that new version looks kind of ugly. I won't resolve this for now. Let me know if I can improve formatting of updated code.

On a side note - I've been resolving conversations about suggestions I've fixed. Is it okay or should I leave resolving convos to maintainers?

Comment (xFrednet) [reply clippy_lints/src/bool_to_int_with_if.rs]: Thank you! The formatting of rustc's if let chains is still open somewhere AFAIK, the advantage Clippy gains from them, is that code inside the `then` and `else` block will be formatted, which wasn't before. I'll have a look at it in the next few days.

> I've been resolving conversations about suggestions I've fixed. Is it okay or should I leave resolving convos to maintainers?

This is different for projects and maintainers. Personally, I (as a reviewer) prefer resolving them myself when I've checked that they've been solved. Simple suggestions can be resolved by either party. But that is just a preference, either way is fine with me. :upside_down_face:

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (jst-r) [reply clippy_lints/src/bool_to_int_with_if.rs]: fixed in [ddc16e2](https://github.com/rust-lang/rust-clippy/pull/9086/commits/ddc16e21f7fd208bee05d4c5df26c3060ae2bd9a)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (jst-r) [reply clippy_lints/src/bool_to_int_with_if.rs]: fixed in [85b9c05](https://github.com/rust-lang/rust-clippy/pull/9086/commits/85b9c053154a449193b6e3dff9e3fcb35365c5f3)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (jst-r) [reply clippy_lints/src/bool_to_int_with_if.rs]: fixed in [a673d6c](https://github.com/rust-lang/rust-clippy/pull/9086/commits/a673d6c8493fe4bc7e04638f247743db8f7946ef)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:38]: I believe that this lint will be part of "1.65.0" due the way that changes are synced back into rust-lang/rust :)
```suggestion
    #[clippy::version = "1.65.0"]
```

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:53]: In this function is a weird mixture of tabs and spaces. That also makes the formatting look weird. I'm surprised that `rustfmt` didn't catch this. Could you replace all of these with spaces?

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:58]: Small formatting nit: If this wasn't let chains which are not yet fully supported by rustfmt, I believe the `&&` would be on the next lines like this:

```rs
if let ExprKind::If(check, then, Some(else_)) = expr.kind
    && let Some(then_lit) = int_literal(then)
    && let Some(else_lit) = int_literal(else_)
    && check_int_literal_equals_val(then_lit, 1)
    && check_int_literal_equals_val(else_lit, 0)
```

Could you adjust this :)

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:104]: Same `&&` formatting NIT as above:

```rs
    if  let ExprKind::Block(block, _) = expr.kind
        && let Block {
            stmts: [],       // Shouldn't lint if statements with side effects
            expr: Some(expr),
            ..
        } = block
        && let ExprKind::Lit(lit) = &expr.kind
        && let LitKind::Int(_, _) = lit.node
```

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:115]:
```rs
    if let ExprKind::Lit(lit) = &expr.kind
       && let LitKind::Int(val, _) = lit.node
       && val == expected_value
```

File: clippy_lints/src/bool_to_int_with_if.rs
Comment (xFrednet) [inline clippy_lints/src/bool_to_int_with_if.rs:13]:
```suggestion
    ///
    /// ### Why is this bad?
```

**Review by xFrednet — APPROVED** [review]
The implementation looks good to me! I have some small formatting nits which should be simple to fix. Then the implementation is ready :tada: 

Before we can merge it, there are two more things. Rust and Clippy use a *No merge commit policy*. This basically means that then only merge commits should be added for PRs. Conflicts have to be resolved with rebases. See [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts). And it would be good if you could squash most commits, so we only have some few meaningful ones left.

To avoid dropping the merge commits first etc. you can squash all commits, including the merge commits, into one. Then you don't have to deal with them separately. You're welcome to reach out, if you need any help with that :)

**Comment by xFrednet** [comment]
Hey @jst-r, the last version looks good so good to me. Now all that needs to be done is a rebase and squashing some commits as well as removing the merge commits. I would suggest squashing everything into one, large commit, as that would also remove the merge commits. Let me know if you need any help with that :upside_down_face:.

**Comment by xFrednet** [comment]
Hey @jst-r, I would love to get this merged, since the lint is a nice addition to Clippy. Since you haven't replied, would it be alright with you, if I squash the commits and merge it? :upside_down_face:

**Comment by jst-r** [comment]
Hi, it would be great if you could merge it. Sorry for dissappearing again, deadlines tied to back to school got the best of me.

**Comment by xFrednet** [comment]
Alright, I'm closing this in favor of https://github.com/rust-lang/rust-clippy/pull/9412. Sadly I couldn't push to this branch since your master is protected. I hope that everything works out and you get credit for the commit. The rebase actually took some time, but the test are all passing now.

> Hi, it would be great if you could merge it. Sorry for dissappearing again, deadlines tied to back to school got the best of me.

No worries, that also happens to me. I hope that you still enjoyed working on Clippy. Thank you for implementing this lint :)

**Pull request closed without merging**
